Remove in-app language pref

nbradbury opened the pull request:
This PR removes the in-app language shared preference, so we now rely on the system's per-app language preference. I also removed the various calls to `setLocale` throughout the app since that's now handled by the system.

Note that for now `LocaleManager.getLanguage()` simply calls `PerAppLocaleManager` to get the current language code. This was to avoid replacing all the calls to `LocaleManager.getLanguage()` throughout the app. I will address that in a subsequent PR because I didn't want to make this one any larger than it is.

Also note that there's still more cleanup to do, but once this PR is merged I think we're good to merge [the feature branch](https://github.com/wordpress-mobile/WordPress-Android/pull/21538) and I can handle the remaining cleanup separately.

nbradbury committed:
First pass at removing language shared preference

nbradbury committed:
Second pass at removing language shared preference

nbradbury committed:
Changed LocaleManager to rely on PerAppLocaleManager for the current language code

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

nbradbury committed:
Removed setLocale

nbradbury committed:
Updated comment

nbradbury committed:
Removed updating the language shared pref

nbradbury committed:
Added TODO

nbradbury committed:
Remove language pref during migration

nbradbury committed:
Merge branch 'feature/per-app-language-prefs' into nick/remove-language-pref

nbradbury committed:
Minor cleanup

nbradbury committed:
Merge branch 'feature/per-app-language-prefs' into nick/remove-language-pref

nbradbury committed:
Merge branch 'feature/per-app-language-prefs' into nick/remove-language-pref

nbradbury committed:
Removed language detection from PersonalizationViewModel, removed isSameLanguage

nbradbury committed:
Updated comment

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/util/LocaleManager.kt:
```diff
@@ -29,37 +19,17 @@ object LocaleManager {
-     * If the user has selected a language other than the device default, return that
-     * language code, else just return the device default language code.
+     * This is simply a wrapper for the per-app language code.
+     * TODO: Remove this and directly call PerAppLocaleManager
```
> Note that for now LocaleManager.getLanguage() simply calls PerAppLocaleManager to get the current language code. This was to avoid replacing all the calls to LocaleManager.getLanguage() throughout the app. I will address that in a subsequent PR because I didn't want to make this one any larger than it is.

Suggestion (💡): Let's consider removing this and directly call `PerAppLocaleManager` now, and before we merge [the feature branch](https://github.com/wordpress-mobile/WordPress-Android/pull/21538) into `trunk`, usages show only 7 placed, and few more via the wrapper, with some classes already being touched by this PR anyway. I think having `LocalManager` and `LocalManagerWrapper` totally free of any language related functionality before merging this whole solution to `trunk` is a good win that is worth pursuing, even if this PR gets a bit more bigger. 🤔

nbradbury replied:
I would really prefer to do this separately. One reason for this is that I'm working a short day today and I'll be AFK until January 3, so I'd prefer to have this feature branch merged before I'm gone. There's not really any user impact for keeping this as-is.

ParaskP7 replied:
Noted Nick, this is purely a suggestion, you know the schedule best and if you do need that merged and going live to users before you return January 3rd, then it makes total sense to me too! 👍

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/util/LocaleManager.kt:
```diff
@@ -116,9 +66,10 @@ object LocaleManager {
+     * TODO: This is a workaround for a bug in API 7, which we no longer support. Investigate removing this.
      */
     @JvmStatic
     fun getSafeLocale(context: Context?): Locale {
```
Praise (❤️): Good find, yea, let's try and remove that, long overdue.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/util/PerAppLocaleManager.kt:
```diff
@@ -31,86 +31,26 @@ class PerAppLocaleManager @Inject constructor(
+                "PerAppLocaleManager: performing migration to AndroidX per-app language prefs"
+            )
+            setCurrentLocaleByLanguageCode(previousLanguage)
+            appPrefsWrapper.removePref(OLD_LOCALE_PREF_KEY_STRING)
```
Question (❓): This prefs removal wouldn't happen for users that are already migrated via #21504 and [this call](https://github.com/wordpress-mobile/WordPress-Android/pull/21504/files#diff-49dc0f74afe58f57b2d6b8640d70df34081db077e00b559ee817fc79e5e0ed60R370) being introduced back then and already live, right? 🤔

nbradbury replied:
That's correct but I believe it's mostly harmless.

ParaskP7 replied:
So, to make sure I understand it fully, those users that are already migrated with the current production version of the app will not have this pref removed, only existing users that haven't yet updated to that version would eventually have this pref removed, when they get this new app update, right? 😊

nbradbury replied:
For those users who have already been migrated, this pref would still be removed because the check for `isNotEmpty`. In other words, we do the migration if the pref exists, which will remove the pref. This does mean that users who were previously migrated would be migrated again, but I don't see any harm in that.

ParaskP7 replied:
Ah cool, totally got it now Nick, thanks! 🙇

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/util/PerAppLocaleManager.kt:
```diff
@@ -155,4 +95,30 @@ class PerAppLocaleManager @Inject constructor(
+        AppCompatDelegate.setApplicationLocales(LocaleListCompat.getEmptyLocaleList())
+    }
+
+    companion object {
```
Question (❓): Why creating the companion object now? 🤔 
Suggestion (💡): Although I kind of understand the need for doing that, I recommend reverting to using those function non-statically, like it was done before, especially after the `LocalManager.getLanguage()` is replaced by `PerAppLocaleManager.getCurrentLocaleLanguageCode()` (which I understand is the only reason this happened in the first place, right? 🤔 

FYI: Btw, I am only questioning and suggesting this because doing that makes this whole solution more (unit) testable.

nbradbury replied:
You are correct that making this static was so `LocalManager.getLanguage()` could access it. I do plan to make it non-static after `LocalManager.getLanguage()` is dropped, but I'd prefer to make that change in a separate PR.

ParaskP7 replied:
Awesome, works for me Nick! 👍

ParaskP7 reviewed (approved):
👋 @nbradbury !

I have reviewed this PR as per the instructions (also tested it), everything works as expected, this is a good clean-up, we're almost there! 🎉

-----

I have left few questions (❓) and a couple of suggestions (💡) for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

-----

`EXTRAS`
1. Question (❓): On #21541 I left a comment about this [val split](https://github.com/wordpress-mobile/WordPress-Android/pull/21541#discussion_r1891781541) and via your reply [here](https://github.com/wordpress-mobile/WordPress-Android/pull/21541#issuecomment-2553913865) I understood that this might also be removed in this PR and prior us merging to `trunk`, thus no point discussing back then. But, if that won't happen, can we discuss that once more, so I undertand why the Java code got transformed to this version of Kotlin code. 🙏 
2. Suggestion (💡): I am not trying to be pedantic here, honestly I don't want to, please take this with a grain of salt; when creating a commit message, please consider providing a reason for a change (commit description), that is, in addition to describing what this commit does, which you do do (commit title). I am only asking that, because it would make me more effective in reviewing the changes as I would understand the why of the change in detail, the reasoning behind that, not only the how ([example](https://github.com/wordpress-mobile/WordPress-Android/pull/21542/commits/3863459df67888835ac387996af64b2672902e35)). 🙏

nbradbury commented:
> please take this with a grain of salt; when creating a commit message, please consider providing a reason for a change (commit description), that is, in addition to describing what this commit does,

That's a very fair comment, I'll do this going forward.

nbradbury committed:
Simplify splitting language display name and code

nbradbury commented:
> 1. can we discuss that once more, so I undertand why the Java code got transformed to this version of Kotlin code. 🙏

Apologies, I dropped the ball on that one. I just corrected it in [6c20c6a](https://github.com/wordpress-mobile/WordPress-Android/pull/21542/commits/6c20c6aeaea472261a6959cbd468dd069b071d7e).

ParaskP7 commented:
> Apologies, I dropped the ball on that one. I just corrected it in [6c20c6a](https://github.com/wordpress-mobile/WordPress-Android/pull/21542/commits/6c20c6aeaea472261a6959cbd468dd069b071d7e).

Oh, awesome, thanks Nick, so all the below got somehow added during the Java-to-Kotlin conversion automagically? 🤔 😨 

```
                      ...toRegex())
                .dropLastWhile { it.isEmpty() }
                .toTypedArray()
```

nbradbury commented:
> so all the below got somehow added during the Java-to-Kotlin conversion automagically?

It did, surprisingly!
